### PR TITLE
Feature/arbitrary state

### PR DESCRIPTION
**Pull request by joelin0**
Creates an arbitrary state given an input complex vector using controlled rotations and phases. Like PR #34 but without using a defgate. Optimizations and other feedback appreciated!

### CODE REVIEW

**Comment by joelin0** [comment]
While this currently works, the gate count grows at an insane rate. I am trying to implement it in another way to reduce this count. Until then, any high level suggestions to reduce gate count are welcome, though in-depth code review is probably too early.

File: grove/unitary_circuit/arbitrary_state.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+def create_arbitrary_state(vector):
+    """
+    This function makes a program that can generate an arbitrary state.
+
+    Uses:
+        - http://140.78.161.123/digital/2016_ismvl_logic_synthesis_quantum_state_generation.pdf
+        - https://arxiv.org/pdf/quant-ph/0407010.pdf
+
+    Given a complex vector a with components a_i (i from 0 to N-1), produce a function
+    that takes in |0> and gives sum_i=1^N a_i/|a| |i>
```
Comment (ncrubin) [inline grove/unitary_circuit/arbitrary_state.py]: Should use the rst format for equations so reads properly from sphinx

**Comment by joelin0** [comment]
@ncrubin It actually turned out that the methods in the utils folder were too inefficient in the number of gates they generated. I'm removing them from this specific PR, since they're not needed for the new way of generating arbitrary states. I'm going to break up the method and clean up documentation for the new file. Sorry for the inconvenience!

File: docs/arbitrary_state.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. automodule:: grove.arbitrary_state.unitary_operator
+    :members:
+    :undoc-members:
+    :show-inheritance:
+
+.. rubric:: References
+
+.. [1] http://140.78.161.123/digital/2016_ismvl_logic_synthesis_quantum_state_generation.pdf
```
Comment (ncrubin) [inline docs/arbitrary_state.rst]: The doc builds shows `(1, 2)` for both papers.

File: grove/arbitrary_state/arbitrary_state.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        # in the phase unification and probability unification steps,
+        # respectively
+        z_thetas = []
+        y_thetas = []
+
+        # will hold updated phases and magnitudes after rotations
+        new_phases = []
+        new_magnitudes = []
+
+        for i in xrange(0, len(phases), 2):
```
Comment (ncrubin) [inline grove/arbitrary_state/arbitrary_state.py]: This section can be more modular to make for easier reading and testing.  You could have an extra function that takes a set of phases and returns two lists `new_phases` and `new_magnitudes`.

File: grove/arbitrary_state/arbitrary_state.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            c = np.sqrt((a ** 2 + b ** 2) / 2.)
+            new_magnitudes.append(c)
+
+        # convert these rotation angles
+        # to those for uncontrolled rotations + CNOTs
+        converted_z_thetas = np.dot(M, z_thetas)
+        converted_y_thetas = np.dot(M, y_thetas)
+
+        # phase unification
+        for j in xrange(len(converted_z_thetas)):
```
Comment (ncrubin) [inline grove/arbitrary_state/arbitrary_state.py]: Make a separate function so you can check correctness of the programs coming out of each phase unification and probability unification step.

File: grove/arbitrary_state/arbitrary_state.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            if converted_z_thetas[j] != 0:
+                # angle is negated in conjugated/reversed circuit
+                reversed_gates.append(RZ(-converted_z_thetas[j], qubits[0]))
+            if step < n - 1:
+                reversed_gates.append(CNOT(qubits[step + rotation_cnots[j]],
+                                           qubits[0]))
+
+        phases = new_phases
+
+        # probability unification
```
Comment (ncrubin) [inline grove/arbitrary_state/arbitrary_state.py]: See above.

File: grove/arbitrary_state/arbitrary_state.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    return M
+
+
+def get_cnot_control_positions(k):
+    """
+    Returns a list of positions for the controls of the CNOTs used when
+    decomposing uniformly controlled rotations, as outlined in
+    arXiv:quant-ph/0407010.
+
+    It is assumed that the target is the zeroth qubit
```
Comment (ncrubin) [inline grove/arbitrary_state/arbitrary_state.py]: zeroth qubit in the list `k` correct? Might want to make the wording a little better.

File: docs/arbitrary_state.rst (hunk 2 above)
Comment (ncrubin) [inline docs/arbitrary_state.rst]: Nevermind.  I see why it show's 1, 2.  Those are the references to that paper.

File: grove/unitary_circuit/utils.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
+"""Utils for generating circuits from unitaries."""
+from pyquil.gates import *
+import numpy as np
+from scipy.linalg import sqrtm
+import pyquil.quil as pq
+
+############### Circuits from single bit unitaries and controlled unitaries ######################
```
Comment (joelin0) [inline grove/unitary_circuit/utils.py]: Code removed in commit 546db05.

File: grove/unitary_circuit/arbitrary_state.py (hunk 1 above)
Comment (joelin0) [inline grove/unitary_circuit/arbitrary_state.py]: Docstring refactored.

File: docs/arbitrary_state.rst [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+\\vert \\Psi \\rangle = \\sum_{i=0}^{N-1}\\frac{a_i}{\\vert \\mathbf{a}\\vert} \\vert i\\rangle
+$$
+
+where \\(\\vert i\\rangle\\) is interpreted by taking \\(i\\) in its binary representation.
+
+This problem is approached in two different ways in this module, and will be described
+in the sections to follow.
+The first is to directly construct a circuit using a sequence of CNOT, rotation, Hadamard, and phase gates,
+that produces the desired state.
+The second is to construct a unitary matrix could be decomposed into different
```
Comment (joelin0) [inline docs/arbitrary_state.rst]: fixed!

File: grove/arbitrary_state/arbitrary_state.py (hunk 6 above)
Comment (joelin0) [inline grove/arbitrary_state/arbitrary_state.py]: `k` is actually an `int`, which is just the number of control qubits. I've updated the docstring; it really doesn't matter what the target qubit is labelled, I just wanted to make it clear that this function assumes the controls are labelled 1, 2, ..., k in order from bottom to top and the indices in the return list are labelled with this convention.

File: grove/arbitrary_state/arbitrary_state.py (hunk 3 above)
Comment (joelin0) [inline grove/arbitrary_state/arbitrary_state.py]: Done (will write tests for it as well).

Comment (joelin0) [inline grove/arbitrary_state/arbitrary_state.py]: Tests are done too.

File: grove/arbitrary_state/arbitrary_state.py (hunk 4 above)
Comment (joelin0) [inline grove/arbitrary_state/arbitrary_state.py]: Done, with tests.

File: grove/arbitrary_state/arbitrary_state.py (hunk 5 above)
Comment (joelin0) [inline grove/arbitrary_state/arbitrary_state.py]: Done, with tests.

**Comment by joelin0** [comment]
@ncrubin I've made revisions as per your code review. Let me know what you think!

File: grove/arbitrary_state/unitary_operator.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+        (1, 0, \\ldots, 0)^T
+
+    :param 1d array state_vector: Normalized vector whose length is at least
+                                  two and a power of two.
+    :return: Unitary operator that encodes state_vector
+    :rtype: 2d array
+    """
+
+    print np.linalg.norm(state_vector)
```
Comment (ncrubin) [inline grove/arbitrary_state/unitary_operator.py]: @harleypatton @joelin0 remove extra print

File: grove/arbitrary_state/unitary_operator.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    assert np.allclose([np.linalg.norm(state_vector)], [1]), \
```
Comment (ncrubin) [inline grove/arbitrary_state/unitary_operator.py]: raise exceptions here and the next assert.  As this is not a "hidden" function a user could still pass in a violating state despite the API docstring specifying valid parameters.

File: grove/arbitrary_state/unitary_operator.py (hunk 9 above)
Comment (joelin0) [inline grove/arbitrary_state/unitary_operator.py]: Done!

File: grove/arbitrary_state/unitary_operator.py (hunk 10 above)
Comment (joelin0) [inline grove/arbitrary_state/unitary_operator.py]: Changed assertions to raising ValueError.

**Comment by joelin0** [comment]
@ncrubin I've made the small changes you've suggested to `unitary_operatory.py`

**Comment by ncrubin** [comment]
Merged!

**Pull request merged**